File: frontend/src/admin/types/event.ts
export interface EventListItem {
  id: number;
  slug: string;
  event_name: string;
  city: string | null;
  country: string | null;
  year: number;
  month: number | null;
  status: string;
  publication_status: "DRAFT" | "SCHEDULED" | "PUBLISHED" | "ARCHIVED";
  is_featured: boolean;
  is_international_milestone: boolean;
  deleted_at: string | null;
}

export interface EventAdminDetail {
  id: number;
  slug: string;
  event_name: string;
  talk_id: number | null;
  session_title: string | null;
  short_description_en: string | null;
  short_description_it: string | null;
  full_description_en: string | null;
  full_description_it: string | null;
  start_date: string | null;
  end_date: string | null;
  year: number;
  month: number | null;
  city: string | null;
  country: string | null;
  continent: string | null;
  latitude: number | null;
  longitude: number | null;
  venue: string | null;
  format: string;
  language: string;
  event_url: string | null;
  slides_url: string | null;
  recording_url: string | null;
  image: string | null;
  status: string;
  sessions_count: number;
  is_featured: boolean;
  is_international_milestone: boolean;
  internal_notes: string | null;
  tags: string[];
  publication_status: "DRAFT" | "SCHEDULED" | "PUBLISHED" | "ARCHIVED";
  publish_at: string | null;
  unpublish_at: string | null;
  published_at: string | null;
  deleted_at: string | null;
  created_at: string;
  updated_at: string;
}

export type EventWritePayload = Omit<
  EventAdminDetail,
  | "id"
  | "tags"
  | "publication_status"
  | "publish_at"
  | "unpublish_at"
  | "published_at"
  | "deleted_at"
  | "created_at"
  | "updated_at"
> & { tag_labels: string[] };

export type EventCreatePayload = Omit<EventWritePayload, "slug"> & { slug: string };
export type EventUpdatePayload = EventWritePayload;

export interface DuplicateCandidate {
  slug: string;
  event_name: string;
  city: string | null;
  year: number;
}

export interface RevisionSummary {
  id: number;
  schema_version: number;
  snapshot: Record<string, unknown>;
  created_at: string;
  created_by_email: string | null;
}

export interface RevisionDiff {
  current: Record<string, unknown>;
  snapshot: Record<string, unknown>;
}

export interface TagSummary {
  slug: string;
  label: string;
}

File: backend/app/models/event.py
from __future__ import annotations

from datetime import date
from typing import TYPE_CHECKING

from sqlalchemy import JSON, Boolean, Date, Float, ForeignKey, Integer, String, Text
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base, TimestampMixin

if TYPE_CHECKING:
    from app.models.talk import Talk


class Event(Base, TimestampMixin):
    """A single speaking engagement (conference, meetup, workshop...)."""

    __tablename__ = "event"

    id: Mapped[int] = mapped_column(primary_key=True)
    slug: Mapped[str] = mapped_column(String(160), unique=True, index=True, nullable=False)

    event_name: Mapped[str] = mapped_column(String(200), nullable=False)
    talk_id: Mapped[int | None] = mapped_column(ForeignKey("talk.id"), nullable=True)
    session_title: Mapped[str | None] = mapped_column(String(255), nullable=True)
    short_description_en: Mapped[str | None] = mapped_column(Text, nullable=True)
    short_description_it: Mapped[str | None] = mapped_column(Text, nullable=True)
    full_description_en: Mapped[str | None] = mapped_column(Text, nullable=True)
    full_description_it: Mapped[str | None] = mapped_column(Text, nullable=True)

    start_date: Mapped[date | None] = mapped_column(Date, nullable=True, index=True)
    end_date: Mapped[date | None] = mapped_column(Date, nullable=True)
    year: Mapped[int] = mapped_column(Integer, nullable=False, index=True)
    # Many real engagements are only known to the month ("September 2026"),
    # without a confirmed exact day. Kept separate from start_date so partial
    # precision can be represented honestly instead of guessing a day.
    month: Mapped[int | None] = mapped_column(Integer, nullable=True)

    city: Mapped[str | None] = mapped_column(String(120), nullable=True, index=True)
    country: Mapped[str | None] = mapped_column(String(120), nullable=True, index=True)
    continent: Mapped[str | None] = mapped_column(String(60), nullable=True, index=True)
    latitude: Mapped[float | None] = mapped_column(Float, nullable=True)
    longitude: Mapped[float | None] = mapped_column(Float, nullable=True)
    venue: Mapped[str | None] = mapped_column(String(200), nullable=True)

    format: Mapped[str] = mapped_column(String(40), nullable=False, default="conference")
    language: Mapped[str] = mapped_column(String(10), nullable=False, default="en")
    topics: Mapped[list[str]] = mapped_column(JSON, nullable=False, default=list)

    event_url: Mapped[str | None] = mapped_column(String(255), nullable=True)
    slides_url: Mapped[str | None] = mapped_column(String(255), nullable=True)
    recording_url: Mapped[str | None] = mapped_column(String(255), nullable=True)
    image: Mapped[str | None] = mapped_column(String(255), nullable=True)

    status: Mapped[str] = mapped_column(String(20), nullable=False, default="confirmed")
    sessions_count: Mapped[int] = mapped_column(Integer, nullable=False, default=1)
    is_featured: Mapped[bool] = mapped_column(Boolean, default=False)
    # A small, deliberately curated subset of "featured" — major geographic
    # expansion moments (first talk in a new country/region) rather than
    # talks with especially notable session content.
    is_international_milestone: Mapped[bool] = mapped_column(Boolean, default=False)

    talk: Mapped[Talk | None] = relationship(back_populates="events")

    @property
    def has_coordinates(self) -> bool:
        return self.latitude is not None and self.longitude is not None
